Fix composite op redispatch for NJT in inference mode

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #134683
* #136566

Prior to this PR, calling `reshape()` under `inference_mode()` would throw a `NotImplementedError`. This is because `inference_mode()` disables autograd key dispatch, incidentally preventing the decomposition of reshape for NJT.

This PR fixes this by redispatching on the `CompositeImplicitAutogradNestedTensor` key whenever a composite implicit op is encountered in `NJT.__torch_dispatch__()`. This fixes reshape and any other composite implicit ops underneath `inference_mode()`.

jbschlosser committed:
Fix composite op redispatch for NJT in inference mode

[ghstack-poisoned]

jbschlosser committed:
Update on "Fix composite op redispatch for NJT in inference mode"


Prior to this PR, calling `reshape()` under `inference_mode()` would throw a `NotImplementedError`. This is because `inference_mode()` disables autograd key dispatch, incidentally preventing the decomposition of reshape for NJT.

This PR fixes this by redispatching on the `CompositeImplicitAutogradNestedTensor` key whenever a composite implicit op is encountered in `NJT.__torch_dispatch__()`. This fixes reshape and any other composite implicit ops underneath `inference_mode()`.

[ghstack-poisoned]

jbschlosser committed:
Update on "Fix composite op redispatch for NJT in inference mode"


Prior to this PR, calling `reshape()` under `inference_mode()` would throw a `NotImplementedError`. This is because `inference_mode()` disables autograd key dispatch, incidentally preventing the decomposition of reshape for NJT.

This PR fixes this by redispatching on the `CompositeImplicitAutogradNestedTensor` key whenever a composite implicit op is encountered in `NJT.__torch_dispatch__()`. This fixes reshape and any other composite implicit ops underneath `inference_mode()`.

[ghstack-poisoned]

albanD reviewed (approved):
Pretty cool!

jbschlosser commented:
@pytorchbot merge -i

jbschlosser commented:
follow-up todo: open an issue for the wrapper subclass + enable reentrant dispatch + torch dispatch mode problem. Proposed fix is to tweak the Python key fallback impl so that it works fine when there is no mode nor subclass passed as an argument.

jbschlosser committed:
Update on "Fix composite op redispatch for NJT in inference mode"


Prior to this PR, calling `reshape()` under `inference_mode()` would throw a `NotImplementedError`. This is because `inference_mode()` disables autograd key dispatch, incidentally preventing the decomposition of reshape for NJT.

This PR fixes this by redispatching on the `CompositeImplicitAutogradNestedTensor` key whenever a composite implicit op is encountered in `NJT.__torch_dispatch__()`. This fixes reshape and any other composite implicit ops underneath `inference_mode()`.

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge

jbschlosser committed:
Update on "Fix composite op redispatch for NJT in inference mode"


Prior to this PR, calling `reshape()` under `inference_mode()` would throw a `NotImplementedError`. This is because `inference_mode()` disables autograd key dispatch, incidentally preventing the decomposition of reshape for NJT.

This PR fixes this by redispatching on the `CompositeImplicitAutogradNestedTensor` key whenever a composite implicit op is encountered in `NJT.__torch_dispatch__()`. This fixes reshape and any other composite implicit ops underneath `inference_mode()`.

[ghstack-poisoned]

jbschlosser committed:
Update on "Fix composite op redispatch for NJT in inference mode"


Prior to this PR, calling `reshape()` under `inference_mode()` would throw a `NotImplementedError`. This is because `inference_mode()` disables autograd key dispatch, incidentally preventing the decomposition of reshape for NJT.

This PR fixes this by redispatching on the `CompositeImplicitAutogradNestedTensor` key whenever a composite implicit op is encountered in `NJT.__torch_dispatch__()`. This fixes reshape and any other composite implicit ops underneath `inference_mode()`.

[ghstack-poisoned]

YuqingJ commented on test/test_nestedtensor.py:
```diff
@@ -7356,6 +7356,38 @@ def g(nt):
+            requires_grad=True,
+        )
+
+        with torch.inference_mode():
```
just checking, this PR doesn't fix the issue when there is a custom TorchDispatchMode right?

soulitzer replied:
that's correct

YuqingJ replied:
Thanks. Maybe we can add a unit case test that cover the custom TorchDispatchMode in test_nestedtensor?

jbschlosser replied:
Sure, I'll add an NJT specific test as well :)

jbschlosser committed:
Update on "Fix composite op redispatch for NJT in inference mode"


Prior to this PR, calling `reshape()` under `inference_mode()` would throw a `NotImplementedError`. This is because `inference_mode()` disables autograd key dispatch, incidentally preventing the decomposition of reshape for NJT.

This PR fixes this by redispatching on the `CompositeImplicitAutogradNestedTensor` key whenever a composite implicit op is encountered in `NJT.__torch_dispatch__()`. This fixes reshape and any other composite implicit ops underneath `inference_mode()`.

[ghstack-poisoned]

jbschlosser committed:
Update on "Fix composite op redispatch for NJT in inference mode"


Prior to this PR, calling `reshape()` under `inference_mode()` would throw a `NotImplementedError`. This is because `inference_mode()` disables autograd key dispatch, incidentally preventing the decomposition of reshape for NJT.

This PR fixes this by redispatching on the `CompositeImplicitAutogradNestedTensor` key whenever a composite implicit op is encountered in `NJT.__torch_dispatch__()`. This fixes reshape and any other composite implicit ops underneath `inference_mode()`.

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge
